### PR TITLE
New API to create an ephemeral node that is session aware

### PR DESCRIPTION
**Pull request by pkuwm**
* Creates an ephemeral node that is session aware.
* If the ephemeral node is created successfully, returns true.
* If session is expired, returns false.
* If connection is timed out or interrupted, exception is thrown.

I will complete the descriptions once the code itself is agreed. We will focus on the code logic in this draft.

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java [hunk 1]
```diff
@@ -568,6 +568,71 @@ public void createEphemeral(final String path, final Object data)
     create(path, data, CreateMode.EPHEMERAL);
   }
 
+  /**
+   * Creates a session aware ephemeral node.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: Please explain what "session aware" means here as well?

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java [hunk 2]
```diff
@@ -568,6 +568,71 @@ public void createEphemeral(final String path, final Object data)
...
+   * @throws ZkInterruptedException
+   *           if operation is interrupted, or a required reconnection gets interrupted
+   * @throws IllegalArgumentException
+   *           if called from anything except the ZooKeeper event thread
+   * @throws ZkException
+   *           if any ZooKeeper exception occurs
+   * @throws RuntimeException
+   *           if any other exception occurs
+   */
+  public boolean createEphemeral(final String path, final Object data, final long sessionId)
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: Lei's comment: 
```
lei-xia 2 minutes ago
Many of the code in this method is duplicated with the one in create() (line 522). Another option is to extend that method to take a sessionID (also wrap the content of that method as a new private method to avoid expose it to be public API)?
```

My reply: I tried to reduce the duplicate code. There are 2 things to consider:
1. Because the input data is an object, we need to serialize it. If so, we need to change the api parameters like:
create(final String path, `byte[] data`, final List<ACL> acl, final CreateMode mode, Callable<String> callable);
And we will need to change create() implementation, whether or not we want to do?
2. Object data needs to be serialized and passed to the final create().
I will show you the change for the above idea and see if we could accept it.

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java (hunk 1 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: Yes, will do.

Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: "Creates an ephemeral node that is session aware.
If the ephemeral node is created successfully, returns true.
If session is expired, returns false.
If connection is timed out or interrupted, exception is thrown."

The description is not very accurate. The gist is that given a session id, we need to make sure the created ephemeral node must be in this session. Otherwise, the creation should fail. 
ZkClient has a auto-reconnect feature. This feature makes this create ephemeral node in this session operation not that easy. 
We need handle session expired, disconnected state change. In additional close(). 

So all in all the following semantic should be true. 

"
given a session ID, the api should return true it can create an ephemeral node in this session.
If met irrecoverable condition, the API should return false.
If met recoverable  condition, the API should retry and create the ephemeral node.
"

Here, irrecoverable condition can be: session expired, new session established, zkclient closed, interrrupted, keeper exceptions such as invalid parameters etc.

Recoverable condition can be: session disconnect. (Note, session can still be reconnected with the same session id potentially later.)

**Comment by pkuwm** [comment]
Lei's comments: https://github.com/pkuwm/helix/commit/f6feb711c36c88994dc2abc6f1c021de59ff2ed3

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java (hunk 1 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: Right, thanks for the meaningful description. This is what I am planning to complete. Once the code itself is agreed by us, I will complete the description.

**Comment by lei-xia** [comment]
Can we have only 1 PR and keep all comments in one place.  It is  a bit confusing here.

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java (hunk 1 above)
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: @lei-xia 's comment:
`
Give a more clear description what does this extra sessionId in parameters mean, such as, this method allows you to specify an expected sessionId, if the current sessionId of underlying ZK connection does not match the given sessionId, the creation will fail ....
`

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java [hunk 3]
```diff
@@ -568,6 +568,71 @@ public void createEphemeral(final String path, final Object data)
...
 
+  /**
+   * Creates a session aware ephemeral node.
+   * If the ephemeral node is created successfully, returns true.
+   * If session is expired, returns false.
+   * If connection is timed out or interrupted, exception is thrown.
+   *
+   * @param path path of the ephemeral node being created
+   * @param data data of the ephemeral node being created
+   * @param sessionId session id of the zookeeper client
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: Lei @lei-xia 
`
This is should be the expected session Id of the ZK connection, not necessarily the sessionID of current ZK Connection, rephrase it to make it less confusing. And also make it clear that if the sessionID of current ZK connection does not match the given (expected) sessionId, the operation will fail.
`

**Comment by pkuwm** [comment]
@dasahcc In this PR, I don’t move the code retryUntilConnect to
createEmphemeNode, instead I reuse retryUntilConnect to minimize the
changes.
